cpp/include/raft/comms/mpi_comms.hpp

```diff
@@ -0,0 +1,290 @@
+    if (MPI_SUCCESS != status) {                                            \
+      int mpi_error_string_lenght = 0;                                      \
+      char mpi_error_string[MPI_MAX_ERROR_STRING];                          \
+      MPI_Error_string(status, mpi_error_string, &mpi_error_string_lenght); \
+      printf("MPI call='%s' at file=%s line=%d failed with %s ", #call,     \
+             __FILE__, __LINE__, mpi_error_string);                         \
+    }                                                                       \
+  } while (0)
+
+namespace raft {
+namespace comms {
+
+static MPI_Datatype get_mpi_datatype(const datatype_t datatype) {
+  switch (datatype) {
+    case datatype_t::CHAR:
+      return MPI_CHAR;
+    case datatype_t::UINT8:
+      return MPI_UNSIGNED_CHAR;
+    case datatype_t::INT32:
+      return MPI_INT;
+    case datatype_t::UINT32:
+      return MPI_UNSIGNED;
+    case datatype_t::INT64:
+      return MPI_LONG_LONG;
+    case datatype_t::UINT64:
+      return MPI_UNSIGNED_LONG_LONG;
+    case datatype_t::FLOAT32:
+      return MPI_FLOAT;
+    case datatype_t::FLOAT64:
+      return MPI_DOUBLE;
+    default:
+      // Execution should never reach here. This takes care of compiler warning.
+      return MPI_DOUBLE;
+  }
+}
+
+static MPI_Op get_mpi_op(const op_t op) {
+  switch (op) {
+    case op_t::SUM:
+      return MPI_SUM;
+    case op_t::PROD:
+      return MPI_PROD;
+    case op_t::MIN:
+      return MPI_MIN;
+    case op_t::MAX:
+      return MPI_MAX;
+    default:
+      // Execution should never reach here. This takes care of compiler warning.
+      return MPI_MAX;
+  }
+}
+
+class mpi_comms : public comms_iface {
+ public:
+  mpi_comms(MPI_Comm comm, const bool owns_mpi_comm)
+    : _owns_mpi_comm(owns_mpi_comm),
+      _mpi_comm(comm),
+      _size(0),
+      _rank(1),
+      _next_request_id(0) {
+    int mpi_is_initialized = 0;
+    MPI_TRY(MPI_Initialized(&mpi_is_initialized));
+    RAFT_EXPECTS(mpi_is_initialized, "ERROR: MPI is not initialized!");
+    MPI_TRY(MPI_Comm_size(_mpi_comm, &_size));
+    MPI_TRY(MPI_Comm_rank(_mpi_comm, &_rank));
+    //get NCCL unique ID at rank 0 and broadcast it to all others
+    ncclUniqueId id;
+    if (0 == _rank) NCCL_TRY(ncclGetUniqueId(&id));
+    MPI_TRY(MPI_Bcast((void*)&id, sizeof(id), MPI_BYTE, 0, _mpi_comm));
+
+    //initializing NCCL
+    NCCL_TRY(ncclCommInitRank(&_nccl_comm, _size, id, _rank));
+  }
+
+  virtual ~mpi_comms() {
+    //finalizing NCCL
+    NCCL_TRY_NO_THROW(ncclCommDestroy(_nccl_comm));
+    if (_owns_mpi_comm) {
+      MPI_TRY_NO_THROW(MPI_Comm_free(&_mpi_comm));
+    }
+  }
+
+  int get_size() const { return _size; }
+
+  int get_rank() const { return _rank; }
+
+  std::unique_ptr<comms_iface> comm_split(int color, int key) const {
+    MPI_Comm new_comm;
+    MPI_TRY(MPI_Comm_split(_mpi_comm, color, key, &new_comm));
+    return std::unique_ptr<comms_iface>(new mpi_comms(new_comm, true));
```

Comment:
Just to double check, this works with code requiring NCCL sub-communicators? (as we are initializing NCCL inside mip_comm constructor involving only the processes in the passed mpi_comm object).

Reply:
Yeah, the new `mpi_comms` instance should create the new nccl comm once constructed.